Add apt option --allow-change-held-packages

staelche opened the pull request:
##### SUMMARY
Adds the APT option --allow-change-held-packages which is needed in case you want to change the version of a package even it is on the hold list (apt-mark showhold). Fixes #65325

##### ISSUE TYPE
Feature Pull Request

##### COMPONENT NAME
apt

##### ADDITIONAL INFORMATION
-

Commit:
Add apt option --allow-change-held-packages

samdoran commented on lib/ansible/modules/apt.py:
```diff
@@ -91,6 +91,11 @@
+    description:
+      - Allows changing the version of a package which is on the apt hold list
+    type: bool
+    default: 'no'
```
```suggestion
    default: 'no'
    version_added: '2.12'
```

samdoran reviewed (changes requested):
Please add integration tests for this new feature and create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment). See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

staelche committed:
Update lib/ansible/modules/apt.py

Co-authored-by: Sam Doran <sdoran@redhat.com>

Commit:
Adds changelog fragment and integration tests

staelche commented:
Sorry for keep you waiting. I needed some time to get used to the integration tests :-)

Commit:
Uncomment accidentally commented test

samdoran commented on test/integration/targets/apt/tasks/apt.yml:
```diff
@@ -437,3 +437,48 @@
+    that:
+      - "not apt_result.changed"
+
+# UNINSTALL
+- name: uninstall hello with apt
+  apt:
+    name: hello
+    state: absent
+    purge: yes
+  register: apt_result
```
The cleanup should be done as a handler. `ansible-tests` runs with `ANSIBLE_FORCE_HANDLERS=True`, so handlers always run even if other tasks fail.

staelche replied:
Ok did not see that. I will change it.

samdoran commented on test/integration/targets/apt/tasks/apt.yml:
```diff
@@ -437,3 +437,48 @@
     state: absent
+
+# https://github.com/ansible/ansible/issues/65325
+- name: install hello with apt (to test allow_change_held_packages option)
+  apt:
+    name: hello
+    state: present
+    allow_change_held_packages: yes
+  register: apt_result
```
I don't see where `hello` is added to the list of held packages before running a task using this option.

samdoran replied:
I believe the general outline of these tests should be:
- install an older version of a package
- add a hold to a package with `apt-mark hold [package]`
- verify package is on the hold list with `apt-mark showhold`
- try to update the package and ensure it did not change
- try to update the package with `allow_change_held_packages=True` and ensure it did change
- try to update the package again with `allow_change_held_packages=True` and ensure it did not change
- remove the package and the hold in a handler

staelche replied:
Yes. You are right. The question I was facing was where to get a old package of something. So I decided to at least to check that I did not introduce any BC break since I was not able to find any integration tests about held packages at all (in the apt part of the integration tests). Found them of course now in the dpkg_selections part and will use the old hello package from s3 they used there.

samdoran commented on test/integration/targets/apt/tasks/apt.yml:
```diff
@@ -437,3 +437,48 @@
+  failed_when: False
+  register: dpkg_result
+
+- name: verify installation of hello
+  assert:
+    that:
+      - "apt_result.changed"
+      - "dpkg_result.rc == 0"
+
+- name: verify apt module outputs
+  assert:
+    that:
+      - "'changed' in apt_result"
```
These two tasks can be combined and should use `changed` test.
```suggestion
- name: verify installation of hello
  assert:
    that:
      - apt_result is changed
      - "dpkg_result.rc == 0"
```

samdoran commented on test/integration/targets/apt/tasks/apt.yml:
```diff
@@ -437,3 +437,48 @@
+- name: verify no change on re-install
+  assert:
+    that:
+      - "not apt_result.changed"
```
```suggestion
      - apt_result is not changed
```

Commit:
Adds enhanced integration tests and handler for cleanup

Commit:
Remove not necessary handler

samdoran reviewed (approved):
Very nice. Tests look good now.

We are currently in feature freeze for the 2.11 release. I will merge this into `devel` once we branch for 2.11.

staelche commented:
Damned. I thought I would get this into the very next release... ;-)

easybe commented on changelogs/73629_allow-change-held-packages.yml:
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - apt - Adds APT option ``--allow-change-held-packages`` as module parameter ``allow_change_held_packages`` to allow APT up- or
+    downgrading a package which is on APTs hold list (https://github.com/ansible/ansible/issues/65325)
```
Missing new line

easybe commented on test/integration/targets/apt/defaults/main.yml:
```diff
@@ -1 +1,2 @@
 apt_foreign_arch: i386
+hello_old_version: 2.6-1
```
Missing new line

easybe commented on test/integration/targets/apt/handlers/main.yml:
```diff
@@ -0,0 +1,5 @@
+- name: remove package hello
+  apt:
+    name: hello
+    state: absent
```
Missing new line

easybe reviewed (commented):
People using `cat` will thank you.

Commit:
Adds newlines to new file

samdoran commented:
/rebuild

ejke commented:
Any chance of merging it and resolving https://github.com/ansible/ansible/issues/65325?

samdoran commented:
This looks good. Let me take a look at the test failures and resolve the merge conflict.

samdoran committed:
Break up function signature into multiple lines

It passed the 160 character width. Change it to multi-line for easier reading
and better diffs in the future.

samdoran committed:
Remove unused import

samdoran committed:
Update version added for new feature

samdoran committed:
Move changelog fragment to correct location

Also remove extra newline

samdoran committed:
Fix yamllint errors
